### ISSUE TITLE
v4 duplicate issue

### ISSUE THREAD

**Issue by wph144** [daegalus/dart-uuid#87]
I have v4 duplication problem and i read https://github.com/Daegalus/dart-uuid/issues/59

My situation is that many different clients each create a v4 uuid and save it to the server db. 

In my real case, 250 duplicates out of 30 million occur
although in my tests, 30 million duplicates do NOT occur at all, 

```
void main(List<String> arguments) {
  final uuidSet = <String>{};
  final uuid = Uuid();
  for (int i = 0; i < 30000000; i++) {
    final newUuid = uuid.v4();
    if (uuidSet.contains(newUuid)) {
      print("duplicated $newUuid");
    } else {
      uuidSet.add(newUuid);
    }
  }
}
```

The difference is that in the test one static Random object creates a random number with a 32 bit seed, 
and in my situation many different Random objects each generate a random number.

It seems that the problem did not occur in the 2.x version, and it seems that the problem occurred in the 3.x version.

I am generating v4 uuid without any option as below in each client
<img width="337" alt="image" src="https://user-images.githubusercontent.com/10491866/226239760-845e6cc7-e7ba-4ec6-a5e9-eadef049d651.png">

**Comment by wonpyohong** [comment]
In order to mimic creating new Random objects for each of multiple devices (clients), we reproduce it as follows:

1. clone this package (https://github.com/Daegalus/dart-uuid.git)
2. find the UuilUtil class and delete 1 line and add 1 line as follows
```
class UuidUtil {
  // static final _random = Random();                       <--------- delete line
  static final _secureRandom = Random.secure();

  static Uint8List mathRNG({int seed = -1}) {
    final _random = Random();                                   <-------- add line
    final b = Uint8List(16);
    final rand = (seed == -1) ? _random : Random(seed);

...
```

3. create main function on new file.
```
import 'package:uuid/uuid.dart';
import 'package:uuid/uuid_util.dart';

void main() {
  int duplicatedCount = 0;
  final uuidSet = <String>{};
  for (int i = 0; i < 1000000; i++) {
    final uuid = Uuid();
    final newUuid = uuid.v4();
    if (uuidSet.contains(newUuid)) {
      // print("duplicated ${newUuid}");
      duplicatedCount++;
    } else {
      uuidSet.add(newUuid);
    }

    if (i % 10000 == 0) {
      print(i);
    }
  }

  print('end. duplicatedCount: $duplicatedCount');
}
```

**Comment by wph144** [comment]
I've created an issue about the problem with random objects in the dart language. https://github.com/dart-lang/language/issues/2934
However, it seems good to increase randomness by restoring shuffle in the uuid package.

**Comment by julemand101** [comment]
Okay, I can see that it might not really help provide your own `seed` to `Random` since it seems always just take 32 bits from it. Might need some further testing...

But it might be that the solution should be going for `Random.secure` if we want actually good random numbers.

**Comment by wph144** [comment]
Can you show an example where you don't get duplicates while using Random.secure? This is my example, but duplication occurs. Is there something I need to modify in UuidUtil?

```
void main() {
  int duplicatedCount = 0;
  final uuidSet = <String>{};
  for (int i = 0; i < 1000000; i++) {
    final uuid = Uuid(options: {'gRng': UuidUtil.cryptoRNG});
    final newUuid = uuid.v4().replaceAll("-", "");

    if (uuidSet.contains(newUuid)) {
      print("duplicated ${newUuid}");
      duplicatedCount++;
    } else {
      uuidSet.add(newUuid);
    }

    if (i % 100000 == 0) {
      print("$i");
    }
  }

  print('end. duplicatedCount: $duplicatedCount');
}
```

**Comment by julemand101** [comment]
@wph144 I changed your program to do 10000000 runs and ran it 3 times without hitting a duplicate.

**Comment by julemand101** [comment]
Made some performance improvements of your program and ran up to 100000000 times:
```dart
import 'dart:collection';

import 'package:uuid/uuid.dart';
import 'package:uuid/uuid_util.dart';

void main() {
  final stopWatch = Stopwatch()..start();
  int duplicatedCount = 0;
  final uuidSet = HashSet<String>();
  for (int i = 0; i < 100000000; i++) {
    final uuid = Uuid(options: {'gRng': UuidUtil.cryptoRNG});
    final newUuid = uuid.v4();

    if (uuidSet.add(newUuid) == false) {
      print("duplicated $newUuid");
      duplicatedCount++;
    }

    if (i % 1000000 == 0) {
      print("$i");
    }
  }

  print('end. duplicatedCount: $duplicatedCount, duration: ${stopWatch.elapsed}');
}
```
Still no collisions:
```
end. duplicatedCount: 0, duration: 0:05:29.887675
```

**Comment by wph144** [comment]
Oh sorry there was a typo in 'gRng' ('grng').
In this case, it is executed with the mathRNG. 
I have confirmed that cryptoRNG is slow but no collides up to 6 million. 
Is the seed of dart's Random.secure() 128bit? I can't find it very well

Anyway, I confirmed that cryptoRNG works fine too. 
I also imagined filling 8 of the 16 arrays in mathRNG with random1 objects and the other 8 with random2 objects.

Thank you for your help.

**Comment by julemand101** [comment]
> Is the seed of dart's Random.secure() 128bit? I can't find it very well

There are no seed since Dart will ask the operating system for a random number each time when you have `Random.secure()` instance. So the performance of this operation is going to be very much dependent on how quick the random generator are on a given platform (including the call into the kernel). E.g. on recent Linux kernels, there have been some major performance improvements when it comes to random generation.

**Comment by wph144** [comment]
I checked 4.0.0-beta2. It's good that String like 'grng' disappears. However, the parameters look a bit complicated at first glance, as you're trying to make them more general so that customRNGs are also possible.

In summary, it looks like the following. I tried other methods such as enum or factory method, but it doesn't seem to be easy.
- math RNG: need seed(int) parameter, fixed implementation
- crypto RNG: no need parameter, fixed implementation
- custom RNG: allows the user to set the number of parameters and parameter types, User should be able to inject implementation

If you want to support customRNG, it would be best to make it into a class like the one below. 
mathRNG and cryptoRNG are inside the package, and for customRNG, the user can create a constructor according to the number and type of parameters and implement it himself. 
However, if user do not read the readme, it is difficult to notice that custom is possible, and it is difficult to notice that two types of MathRng and CryptoRng are supported by default. (It is easy to notice if it is made with enum, but it is difficult to process parameters for custom)
It is just draft and just suggestion. It is up to you.

```dart
abstract class RNG {
  Uint8List generate();
}

class MathRNG implements RNG {
  static final _random = Random();
  int seed;

  MathRNG({this.seed = -1});

  @override
  Uint8List generate() {
    final b = Uint8List(16);
    final rand = (seed == -1) ? _random : Random(seed);

    for (var i = 0; i < 16; i++) {
      b[i] = rand.nextInt(256);
    }

    return b;
  }
}

class CryptoRNG implements RNG {
  static final _secureRandom = Random.secure();

  @override
  Uint8List generate() {
    final b = Uint8List(16);

    for (var i = 0; i < 16; i++) {
      b[i] = _secureRandom.nextInt(256);
    }

    return b;
  }
}

// user(developer)'s app
class CustomRNG implements RNG {
  // parameters
  
  // constructor if you need parameters

  @override
  Uint8List generate() {
    final b = Uint8List(16);

    // implementation

    return b;
  }
}
```dart

**Comment by daegalus** [comment]
I will consider this as this is a good suggestion. And a breaking change like this is best to make now before I make 4.0 final, especially since I'm forcing GlobalOptions as a breaking change.

**Comment by daegalus** [comment]
@wph144 give 4.0.0-beta3 a shot, I changed to using RNG classes very much like your suggestion. Let me know if that works for you.

**Comment by wph144** [comment]
@daegalus I checked it. I created PR. https://github.com/daegalus/dart-uuid/pull/91
And you haven't removed the UuidUtil stuff from the docs yet, right? I was going to do it, but I had to fix the html too, so I gave up.

**Comment by daegalus** [comment]
> @daegalus I checked it. I created PR. https://github.com/daegalus/dart-uuid/pull/91
> And you haven't removed the UuidUtil stuff from the docs yet, right? I was going to do it, but I had to fix the html too, so I gave up.

I did generate new docs, but I forgot that it doesn't cleanup removed things. I'll fix it in a few hours. Your PR looks good. I'm going to merge it in once the tests pass.

**Comment by daegalus** [comment]
@wph144 all set, your PR is merged, and i released 4.0.0-beta3-1 with your fixes

**Comment by wph144** [comment]
Thank you for your fast work @daegalus 
I will close this issue
